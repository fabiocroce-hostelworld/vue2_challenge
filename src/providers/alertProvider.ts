import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
