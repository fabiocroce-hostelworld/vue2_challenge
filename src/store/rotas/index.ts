import { getRotaServices } from "@/api/clavaAPI"
import { IRotas } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_ROTAS: "GET_ROTAS"
  },
  mutations: {
    STORE_ROTAS: "STORE_ROTAS",
    STORE_ROTA: "STORE_ROTA"
  },
  actions: {
    FETCH_ROTA: "fetchRota",
    GENERATE_ROTA: "generateRota"
  }
}

export const rotasTypes = addNamespacePrefix(types, "rotas")

type RotasState = {
  rotas: IRotasList
}

export default {
  namespaced: true,
  state: (): RotasState => ({
    rotas: [] as IRotasList
  }),
  getters: {
    [types.getters.GET_ROTAS](state: RotasState) {
      return state.rotas
    }
  },
  actions: {
    [types.actions.FETCH_ROTA]: async ({ commit }: { commit: Function }, rotaId: number) => {
      const rota = await getRotaServices().fetchRota(rotaId)

      commit(types.mutations.STORE_ROTA, rota)
    },
    [types.actions.GENERATE_ROTA]: async ({ dispatch }: { dispatch: Function }, date: string) => {
      const rotaId = await getRotaServices().generateRota(date)

      await dispatch(types.actions.FETCH_ROTA, rotaId)
    }
  },
  mutations: {
    [types.mutations.STORE_ROTAS]: (state: RotasState, rotas: IRotasList) => {
      state.rotas = rotas
    },
    [types.mutations.STORE_ROTA]: (state: RotasState, rota: IRotas) => {
      state.rotas.push(rota)
    }
  }
}
