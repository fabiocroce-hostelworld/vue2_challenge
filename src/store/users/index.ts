import { IUsersList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_USERS: "getUsers",
    GET_MAPPED_USERS: "getMappedUsers",
    GET_USER_NAME: "getUserName"
  },
  mutations: {
    STORE_USERS: "STORE_USERS"
  },
  actions: {
    FETCH_USERS: "fetchUsers"
  }
}

export const usersTypes = addNamespacePrefix(types, "users")

type UsersState = {
  users: IUsersList
}

export default {
  namespaced: true,
  state: (): UsersState => ({
    users: [] as IUsersList
  }),
  getters: {
    [types.getters.GET_USERS](state: UsersState) {
      return state.users
    },
    [types.getters.GET_MAPPED_USERS](state: UsersState) {
      return Object.assign({}, ...state.users.map(user => ({ [user.userId]: user.user })))
    },
    [types.getters.GET_USER_NAME](_: UsersState, getters: { [x: string]: IUsersList }) {
      return (userId: number) => getters[types.getters.GET_MAPPED_USERS][userId]
    }
  },
  mutations: {
    [types.mutations.STORE_USERS]: (state: UsersState, users: IUsersList) => {
      state.users = users
    }
  }
}
