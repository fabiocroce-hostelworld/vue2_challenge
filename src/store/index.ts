import { getRotasUsersServices } from "@/api/clavaAPI"
import Vue from "vue"
import Vuex, { Store } from "vuex"
import rotas, { rotasTypes } from "@/store/rotas"
import users, { usersTypes } from "@/store/users"

Vue.use(Vuex)

export const rootTypes = {
  actions: {
    FETCH_ROTAS_AND_USERS: "fetchRotasAndUsers"
  }
}

export { rotasTypes, usersTypes }

export const storeConfig = {
  modules: {
    rotas,
    users
  },
  actions: {
    [rootTypes.actions.FETCH_ROTAS_AND_USERS]: async ({ commit }: { commit: Function }) => {
      const { users, rotas } = await getRotasUsersServices().fetchUsersAndRotas()

      commit(rotasTypes.mutations.STORE_ROTAS, rotas)
      commit(usersTypes.mutations.STORE_USERS, users)
    }
  }
}

export default new Store(storeConfig)
