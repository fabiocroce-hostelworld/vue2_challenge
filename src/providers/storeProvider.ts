import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
