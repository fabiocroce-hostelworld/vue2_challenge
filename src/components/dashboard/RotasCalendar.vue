<template>
  <v-sheet height="500">
    <v-calendar
      ref="calendar"
      :now="startDate"
      :start="startDate"
      :value="startDate"
      :events="selectedEvents"
      color="primary"
      type="month"
      :event-color="getEventColor"
    />
  </v-sheet>
</template>

<script lang="ts">
import { usersTypes } from "@/store"
import { computed, defineComponent, PropType, reactive } from "@vue/composition-api"
import { ShiftTypes } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { rotasTypes } from "@/store"
import { useStore, StoreProvider } from "@/providers/storeProvider"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"

interface Props {
  selectedUsers: Set<number>
}

export default defineComponent<Props>({
  name: "RotasCalendar",
  props: {
    selectedUsers: {
      type: Set as PropType<Set<number>>,
      required: true
    }
  },
  setup(props) {
    const { store, reactiveGetter } = useStore() as StoreProvider
    const rotas = reactiveGetter<IRotasList>(rotasTypes.getters.GET_ROTAS)
    const reactiveUsers = reactive(props)

    const getUserNameById = (userId: number) => store.getters[usersTypes.getters.GET_USER_NAME](userId)

    const startDate = computed(() => rotas.value[0]?.period.startDate)
    const events = computed(() => {
      return rotas.value.flatMap(thisRota =>
        thisRota.rota.map(({ date, userId, type }) => ({
          userId,
          name: getUserNameById(userId),
          ...(type === ShiftTypes.MORNING
            ? {
                start: `${date} 09:00`,
                end: `${date} 12:00`,
                color: "secondary"
              }
            : {
                start: `${date} 12:00`,
                end: `${date} 17:00`,
                color: "primary"
              })
        }))
      )
    })

    const selectedEvents = computed(() => events.value.filter(rota => reactiveUsers.selectedUsers.has(rota.userId)))

    return {
      startDate,
      selectedEvents,
      /* istanbul ignore next */
      getEventColor(event: { color: string }) {
        return event.color
      },
      rotas
    }
  }
})
</script>
