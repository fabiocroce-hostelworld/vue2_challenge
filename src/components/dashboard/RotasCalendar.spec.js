import Vuex from "vuex"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import { storeConfig, rotasTypes, usersTypes } from "@/store"
import RotasCalendar from "@/components/dashboard/RotasCalendar"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"
import { setupStoreProvider, storeRef } from "@/providers/storeProvider"

describe("RotasCalendar - loaded rotas", () => {
  let input
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()
    store = new Vuex.Store(storeConfig)

    store.commit(rotasTypes.mutations.STORE_ROTAS, rotaUsersMock.rotas)
    store.commit(usersTypes.mutations.STORE_USERS, rotaUsersMock.users)

    input = {
      localVue,
      provide: {
        [storeRef]: setupStoreProvider(store)
      }
    }
  })

  it("Snapshot test", () => {
    input.propsData = {
      selectedUsers: new Set([2, 3])
    }
    const wrapper = shallowMount(RotasCalendar, input)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("expects events to be empty if selectedUsers has no user ids", () => {
    input.propsData = {
      selectedUsers: new Set()
    }
    const wrapper = shallowMount(RotasCalendar, input)

    const calendar = wrapper.find("v-calendar-stub")
    const props = calendar.props()

    expect(props.now).toBe("2020-11-01")
    expect(props.events).toEqual([])
  })

  it("expects events not to be empty if selectedUsers has a userID (Titan Cox)", () => {
    input.propsData = {
      selectedUsers: new Set([5])
    }
    const wrapper = shallowMount(RotasCalendar, input)

    const calendar = wrapper.find("v-calendar-stub")
    const props = calendar.props()

    expect(props.now).toBe("2020-11-01")
    expect(props.events).toEqual([
      { color: "secondary", end: "2020-11-05 12:00", name: "Titan Cox", start: "2020-11-05 09:00", userId: 5 },
      { color: "primary", end: "2020-11-14 17:00", name: "Titan Cox", start: "2020-11-14 12:00", userId: 5 },
      { color: "secondary", end: "2020-11-16 12:00", name: "Titan Cox", start: "2020-11-16 09:00", userId: 5 }
    ])
  })
})

describe("RotasCalendar - no loaded rotas", () => {
  it("expects events to be empty if rotas are not loaded", () => {
    const localVue = createLocalVue()
    const store = new Vuex.Store(storeConfig)
    const input = {
      localVue,
      provide: {
        [storeRef]: setupStoreProvider(store)
      }
    }
    input.propsData = {
      selectedUsers: new Set()
    }
    const wrapper = shallowMount(RotasCalendar, input)

    const calendar = wrapper.find("v-calendar-stub")
    const props = calendar.props()

    expect(props.now).toBe(undefined)
    expect(props.events).toEqual([])
  })
})
