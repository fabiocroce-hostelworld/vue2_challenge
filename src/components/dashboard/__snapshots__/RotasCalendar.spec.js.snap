// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotasCalendar - loaded rotas Snapshot test 1`] = `
<v-sheet-stub height="500" tag="div">
  <v-calendar-stub color="primary" now="2020-11-01" start="2020-11-01" weekdays="0,1,2,3,4,5,6" shortweekdays="true" events="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" eventstart="start" eventend="end" eventtimed="timed" eventcategory="category" eventheight="20" eventcolor="function () { [native code] }" eventtextcolor="white" eventname="name" eventoverlapthreshold="60" eventoverlapmode="stack" eventmore="true" eventmoretext="$vuetify.calendar.moreEvents" eventmarginbottom="1" type="month" value="2020-11-01" localefirstdayofyear="0" minweeks="1" shortmonths="true" showmonthonfirst="true" maxdays="7" shortintervals="true" intervalheight="48" intervalwidth="60" intervalminutes="60" firstinterval="0" intervalcount="24" categories="" categoryforinvalid="" categorydays="1" class="calendar"></v-calendar-stub>
</v-sheet-stub>
`;
